pkg/db/src/implementations/version-1.2/books.ts

```diff
@@ -1,95 +1,102 @@
-import { concat, from, map, Observable, switchMap, tap } from "rxjs";
+import { concat, from, Observable, switchMap, tap } from "rxjs";
 
-import { debug } from "@librocco/shared";
+import { debug, wrapIter, PluginManager } from "@librocco/shared";
 
 import { BookEntry, BooksInterface, DatabaseInterface } from "@/types";
 
 import { newChangesStream, unwrapDocs } from "@/utils/pouchdb";
 
 class Books implements BooksInterface {
 	#db: DatabaseInterface;
+	#pluginManagerInstance: PluginManager;
 
-	constructor(db: DatabaseInterface) {
+	constructor(db: DatabaseInterface, pluginManagerInstance?: PluginManager) {
 		this.#db = db;
+		this.#pluginManagerInstance = pluginManagerInstance || new PluginManager();
 	}
 
 	async upsert(bookEntries: BookEntry[]): Promise<void> {
-		await Promise.all(
-			bookEntries.map((b) => {
-				return new Promise<void>((resolve, reject) => {
-					const bookEntry = { ...b, _id: `books/${b.isbn}` };
-					this.#db._pouch
-						.put(bookEntry)
-						.then(() => resolve())
-						.catch((err) => {
-							// eslint-disable-next-line @typescript-eslint/no-explicit-any
-							if ((err as any).status !== 409) reject(err);
-
-							this.#db._pouch.get(bookEntry._id).then((bookDoc) => {
-								this.#db._pouch.put({ ...bookDoc, ...bookEntry }).then(() => {
-									resolve();
-								});
-							});
-						});
-				});
-			})
-		);
+		const bookIds = bookEntries.map((b) => `books/${b.isbn}`);
+
+		const docs = await this.#db._pouch.allDocs({ keys: bookIds, include_docs: true }).then(({ rows }) => {
+			// rows => contain _id and _rev
+			// Merge result with bookEntries
+			const mergedArray = bookEntries.map((b) => {
+				// Find the corresponding doc in result
+				const match = rows.find((row) => row.id === `books/${b.isbn}`);
+				// If a match is found, return the merged object
+				if (match && match.doc) {
+					return {
+						...match.doc,
+						_rev: match.value.rev,
+						...b
+					};
+				}
```

Comment:
You can zip this...the revs will be there in the same order as the requested keys with `undefined` where the `_rev` couldn't be found